core/assay.js

```diff
@@ -0,0 +1,129 @@
+import harden from '@agoric/harden';
+
+import { insist } from '../util/insist';
+import { mustBeSameStructure, mustBeComparable } from '../util/sameStructure';
+
+// This module treats labels as black boxes. It is not aware
+// of issuers, and so can handle labels whose issuers are merely
+// presences of remote issuers.
+
+// `makeAssayMaker` takes in a 'strategy' object which defines set
+// operations for a particular kind of assay. For instance, for the
+// natAssay, these operations are arithmetic.
+
+// Return an assay, which makes amounts, validates amounts, and
+// provides set operations over amounts. An amount is a pass-by-copy
+// description of some set of erights. An amount has a label and a
+// quantity. All amounts made by the same assay have the same label
+// but differ in quantity.
+//
+// An assay is pass-by-presence, but is not designed to be usefully
+// passed. Rather, we expect each vat that needs to operate on amounts
+// will have its own local assay to do so.
+
+function makeAssay(label, strategy) {
+  mustBeComparable(label);
+
+  // The brand represents recognition of the amount as authorized.
+  const brand = new WeakSet();
+
+  const assay = harden({
+    getLabel() {
+      return label;
+    },
+
+    getStrategy() {
+      return strategy;
+    },
+
+    // Given the raw quantity that this kind of amount would label, return
+    // an amount so labeling that quantity.
+    make(allegedQuantity) {
+      strategy.insistKind(allegedQuantity);
```

Comment:
This is now self consistent, so LGTM for this PR. However, I think the opposite consistent stance, that the strategy contribute a quantity coercing behavior, may be preferable. Since we have no current need for such a strategy-based coercion, we can mull this over non-urgently.